### PR TITLE
[CDAP-20874] Added API handler functions (scan and get) for LRO

### PR DESCRIPTION
**Pull request by adrikagupta**
- Added scanOperations and getOperationRun API handler functions in OperationHttpHandler.java
- Added getOperationRun() in OperationLifecycleManager.java
- Added unit tests in the OperationLifecycleManagerTest.java

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java [hunk 1]
```diff
@@ -98,4 +167,96 @@ public void failOperation(FullHttpRequest request, HttpResponder responder,
...
+    if (filterStr != null && !filterStr.isEmpty()) {
+      OperationRunFilter operationRunFilter = getFilter(filterStr);
+      builder.setFilter(operationRunFilter);
+    }
+    return builder.build();
+  }
+
+  // TODO[CDAP-20895] : Add unit tests for extracting OperationRunFilter from filter string
+  private OperationRunFilter getFilter(String filterStr) throws IllegalArgumentException {
+    Map<String, String> filterKeyValMap = parseKeyValStr(filterStr, FILTER_SPLITTER);
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: nit make it immutable map

**Review by samdgupi — APPROVED** [review]

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java (hunk 1 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: Right, done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java [hunk 2]
```diff
@@ -16,54 +16,112 @@
...
+  @Inject
+  OperationHttpHandler(CConfiguration cConf, OperationLifecycleManager operationLifecycleManager)
+      throws Exception {
+    this.cConf = cConf;
+    this.batchSize = this.cConf.getInt(AppFabric.STREAMING_BATCH_SIZE);
+    this.operationLifecycleManager = operationLifecycleManager;
+    this.featureFlagsProvider = new DefaultFeatureFlagsProvider(cConf);
   }
 
+  // TODO[CDAP-20881] :  Add RBAC check
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: (CDAP-20881)

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java [hunk 3]
```diff
@@ -16,54 +16,112 @@
...
-  
-  OperationHttpHandler() {
+  private final OperationLifecycleManager operationLifecycleManager;
+  private final int batchSize;
+  public static final String OPERATIONS_LIST_PAGINATED_KEY = "operations";
+
+  @Inject
+  OperationHttpHandler(CConfiguration cConf, OperationLifecycleManager operationLifecycleManager)
+      throws Exception {
+    this.cConf = cConf;
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: remove field `cConf` as it is only used inside constructor

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java [hunk 4]
```diff
@@ -16,54 +16,112 @@
...
   private static final Gson GSON = new Gson();
-  
-  OperationHttpHandler() {
+  private final OperationLifecycleManager operationLifecycleManager;
+  private final int batchSize;
+  public static final String OPERATIONS_LIST_PAGINATED_KEY = "operations";
+
+  @Inject
+  OperationHttpHandler(CConfiguration cConf, OperationLifecycleManager operationLifecycleManager)
+      throws Exception {
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: remove `exception`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java [hunk 5]
```diff
@@ -98,4 +167,96 @@ public void failOperation(FullHttpRequest request, HttpResponder responder,
...
+
+  private NamespaceId validateNamespaceId(String namespaceId) throws BadRequestException {
+    try {
+      return new NamespaceId(namespaceId);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(e.getMessage(), e);
+    }
+  }
+
+  /** throws {@link ForbiddenException} if the feature is disabled */
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: do not need this comment

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java [hunk 6]
```diff
@@ -17,6 +17,8 @@
 package io.cdap.cdap.internal.operation;
 
 import com.google.inject.Inject;
+
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java]: remove this empty line

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java [hunk 7]
```diff
@@ -78,6 +80,27 @@ public boolean scanOperations(ScanOperationRunsRequest request, int txBatchSize,
...
+   /**
+ * Retrieves details of an operation run identified by the provided {@code OperationRunId}.
+ *
+ * @param runId The unique identifier for the operation run.
+ * @return An {@code OperationRunDetail} object containing information about the specified operation run.
+ * @throws OperationRunNotFoundException If the specified operation run is not found.
+ */
+  public OperationRunDetail getOperationRun(OperationRunId runId)
+      throws IOException, OperationRunNotFoundException {
+    OperationRunDetail operationRunDetail =
```
Comment (samdgupi) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java]: inline variable i.e return from here directly no need to create a valibale

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/operation/OperationLifecycleManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+import org.junit.rules.TemporaryFolder;
+import org.mockito.Mockito;
+
+public class OperationLifecycleManagerTest extends OperationTestBase {
+  protected static TransactionRunner transactionRunner;
+  private static final String testNamespace = "test";
+  private static OperationLifecycleManager operationLifecycleManager;
+  private static int batchSize;
+
+  @ClassRule public static final TemporaryFolder TEMP_FOLDER = new TemporaryFolder();
```
Comment (samdgupi) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/operation/OperationLifecycleManagerTest.java]: move @ClassRule to previous line

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/operation/OperationLifecycleManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    TransactionRunners.run(
+        transactionRunner,
+        context -> {
+          List<OperationRunDetail> gotRuns = new ArrayList<>();
+          List<OperationRunDetail> expectedRuns;
+          ScanOperationRunsRequest request;
+
+          // verify the scan without filters picks all runs for testNamespace
+          request = ScanOperationRunsRequest.builder().setNamespace(testNamespace).build();
+          operationLifecycleManager.scanOperations(request, batchSize, d -> gotRuns.add(d));
```
Comment (samdgupi) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/operation/OperationLifecycleManagerTest.java]: replace `d -> gotRuns.add(d)` to `gitRuns:add`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java (hunk 2 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: Done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java (hunk 3 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java (hunk 4 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java (hunk 5 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/OperationHttpHandler.java]: done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java (hunk 6 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java]: done

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java (hunk 7 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationLifecycleManager.java]: done

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/operation/OperationLifecycleManagerTest.java (hunk 8 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/operation/OperationLifecycleManagerTest.java]: done

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/operation/OperationLifecycleManagerTest.java (hunk 9 above)
Comment (adrikagupta) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/operation/OperationLifecycleManagerTest.java]: done

**Pull request merged**